// @starknet-io/get-starknet-wallet-standard/dist/chains.mjs
import { Hex } from 'ox';

const STARKNET_CHAIN_PREFIX = "starknet:";
class StarknetChainError extends Error {
  constructor(message) {
    super(message);
    this.name = "StarknetChainError";
  }
}
const WELL_KNOWN_STARKNET_CHAINS = [
  "starknet:0x534e5f4d41494e",
  // mainnet
  "starknet:0x534e5f5345504f4c4941"
  // sepolia
];
function isStarknetChain(chain) {
  const parts = chain.split(":");
  if (parts.length !== 2 || parts[0] !== "starknet") {
    return false;
  }
  return Hex.validate(parts[1]);
}
function formatStarknetChainId(chainId) {
  if (!Hex.validate(chainId)) {
    throw new StarknetChainError(`Invalid Starknet chain id: ${chainId}`);
  }
  return `${STARKNET_CHAIN_PREFIX}${chainId}`;
}
function getStarknetChainId(chain) {
  if (!isStarknetChain(chain)) {
    throw new StarknetChainError(`Invalid Starknet chain: ${chain}`);
  }
  return chain.slice(STARKNET_CHAIN_PREFIX.length);
}

export { STARKNET_CHAIN_PREFIX, StarknetChainError, WELL_KNOWN_STARKNET_CHAINS, formatStarknetChainId, getStarknetChainId, isStarknetChain };
